<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Tutorial · Crispulator.jl</title><script data-outdated-warner src="../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.045/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.13.24/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL=".."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../assets/documenter.js"></script><script src="../siteinfo.js"></script><script src="../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><a class="docs-logo" href="../"><img src="../assets/logo.png" alt="Crispulator.jl logo"/></a><div class="docs-package-name"><span class="docs-autofit"><a href="../">Crispulator.jl</a></span></div><form class="docs-search" action="../search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li><a class="tocitem" href="../">Home</a></li><li class="is-active"><a class="tocitem" href>Tutorial</a><ul class="internal"><li><a class="tocitem" href="#Graphical-overview"><span>Graphical overview</span></a></li><li><a class="tocitem" href="#Getting-started"><span>Getting started</span></a></li><li><a class="tocitem" href="#Running-simulation"><span>Running simulation</span></a></li><li><a class="tocitem" href="#Output"><span>Output</span></a></li></ul></li><li><a class="tocitem" href="../custom/">Custom Simulations</a></li><li><a class="tocitem" href="../internals/">Crispulator Internals</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>Tutorial</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Tutorial</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/tlnagy/Crispulator.jl/blob/master/docs/src/tutorial.md" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="Tutorial"><a class="docs-heading-anchor" href="#Tutorial">Tutorial</a><a id="Tutorial-1"></a><a class="docs-heading-anchor-permalink" href="#Tutorial" title="Permalink"></a></h1><p>The easiest way to use Crispulator is via the command line config file. If this format is too constraining, the <a href="../custom/#Custom-Simulations">Custom Simulations</a> has a detailed walk- through of writing a custom simulation where each step can be modified according to need.</p><h2 id="Graphical-overview"><a class="docs-heading-anchor" href="#Graphical-overview">Graphical overview</a><a id="Graphical-overview-1"></a><a class="docs-heading-anchor-permalink" href="#Graphical-overview" title="Permalink"></a></h2><p>The simulation is laid out in the following manner:</p><p><img src="../assets/overview.png" alt/></p><h2 id="Getting-started"><a class="docs-heading-anchor" href="#Getting-started">Getting started</a><a id="Getting-started-1"></a><a class="docs-heading-anchor-permalink" href="#Getting-started" title="Permalink"></a></h2><p>First, navigate to the Crispulator directory.</p><div class="admonition is-success"><header class="admonition-header">Tip</header><div class="admonition-body"><p>You can find the directory by running</p><pre><code class="language-sh hljs">$ julia -e &#39;import Crispulator; println(normpath(pathof(Crispulator), &quot;..&quot;, &quot;..&quot;))&#39;</code></pre></div></div><p>There should be a <a href="https://learnxinyminutes.com/docs/yaml/">YAML file</a> called <code>example_config.yml</code>. Open this is in a text editor and it should look like this</p><pre><code class="language-yaml hljs"># This is an example configuration file. Whitespace is important.

# Settings pertaining to the library design
library:
    genome:
        num-genes: 500
        num-guides-per-gene: 5
        frac-increasing-genes: 0.02 # fraction of genes with a positive phenotype
        frac-decreasing-genes: 0.1 # fraction of genes with a negative phenotype

    guides:
        crispr-type: CRISPRn # either CRISPRi or CRISPRn
        frac-high-quality: 0.9 # fraction of high quality guides
        mean-high-quality-kd: 0.85 # mean knockdown by a high quality guide (CRISPRi only)

screen:
    type: facs # either facs or growth
    num-runs: 10 # how many independent runs

    representation: # integer value, how much larger are samples than the library
        - transfection: 100
        - selection: 100
        - sequencing: 100

# screen-type specific parameters

    bin-size: 0.25 # size of tail to sample from, must be between 0 and 0.5 (FACS only)
    std-noise: 1 # (FACS only)
    num-bottlenecks: 10 # (Growth only)</code></pre><p>This gives access to most dials in the simulation, if something is missing than see <a href="../custom/#Custom-Simulations">Custom Simulations</a>.</p><p>Now, lets remove all genes that have a positive phenotype by changing line 8 to 0.0:</p><pre><code class="language-yaml hljs">        frac-increasing-genes: 0.0 # fraction of genes with a positive phenotype</code></pre><h2 id="Running-simulation"><a class="docs-heading-anchor" href="#Running-simulation">Running simulation</a><a id="Running-simulation-1"></a><a class="docs-heading-anchor-permalink" href="#Running-simulation" title="Permalink"></a></h2><p>Now, we can actually run the code by executing the following command</p><pre><code class="language-julia hljs">julia run.jl config example_config.yml test_output</code></pre><div class="admonition is-success"><header class="admonition-header">Tip</header><div class="admonition-body"><p>Here <code>config</code> tells <code>CRISPulator</code> to use the provided config <code>example_config.yml</code> and <code>test_output</code> is the directory where the results will be saved. This directory will be created if it doesn&#39;t exist.</p></div></div><p>The output should look like</p><pre class="documenter-example-output"><code class="nohighlight hljs ansi">[ Info: Activating simulation environment
  Activating project at `~/work/Crispulator.jl/Crispulator.jl`
[ Info: Instantiating environment
    Updating registry at `~/.julia/registries/General.toml`
    Updating `~/work/Crispulator.jl/Crispulator.jl/Project.toml`
  [c7e460c6] + ArgParse v1.1.4
  [336ed68f] + CSV v0.10.7
  [a2cac450] + ColorBrewer v0.4.0
  [861a8166] + Combinatorics v1.0.2
  [a81c6b42] + Compose v0.9.4
  [a93c6f00] + DataFrames v1.4.3
  [864edb3b] + DataStructures v0.18.13
  [31c24e10] + Distributions v0.25.79
⌅ [ffbed154] + DocStringExtensions v0.8.6
  [c91e804a] + Gadfly v1.3.4
  [09f84164] + HypothesisTests v0.10.11
  [c8e1da08] + IterTools v1.4.0
  [442fdcdd] + Measures v0.3.2
  [2913bbd2] + StatsBase v0.33.21
  [ddb6d928] + YAML v0.4.8
    Updating `~/work/Crispulator.jl/Crispulator.jl/Manifest.toml`
  [621f4979] + AbstractFFTs v1.2.1
  [79e6a3ab] + Adapt v3.4.0
  [c7e460c6] + ArgParse v1.1.4
  [13072b0f] + AxisAlgorithms v1.0.1
  [336ed68f] + CSV v0.10.7
  [49dc2e85] + Calculus v0.5.1
  [324d7699] + CategoricalArrays v0.10.7
  [d360d2e6] + ChainRulesCore v1.15.6
  [9e997f8a] + ChangesOfVariables v0.1.4
  [944b1d66] + CodecZlib v0.7.0
  [a2cac450] + ColorBrewer v0.4.0
  [3da002f7] + ColorTypes v0.11.4
  [5ae59095] + Colors v0.12.8
  [861a8166] + Combinatorics v1.0.2
  [38540f10] + CommonSolve v0.2.3
  [34da2185] + Compat v4.4.0
  [a81c6b42] + Compose v0.9.4
  [187b0558] + ConstructionBase v1.4.1
⌅ [d38c429a] + Contour v0.5.7
  [7ad07ef1] + CoupledFields v0.2.0
  [a8cc5b0e] + Crayons v4.1.1
  [9a962f9c] + DataAPI v1.13.0
  [a93c6f00] + DataFrames v1.4.3
  [864edb3b] + DataStructures v0.18.13
  [e2d170a0] + DataValueInterfaces v1.0.0
  [b429d917] + DensityInterface v0.4.0
  [b4f34e82] + Distances v0.10.7
  [31c24e10] + Distributions v0.25.79
⌅ [ffbed154] + DocStringExtensions v0.8.6
  [fa6b7ba4] + DualNumbers v0.6.8
  [7a1cc6ca] + FFTW v1.5.0
  [48062228] + FilePathsBase v0.9.20
  [1a297f60] + FillArrays v0.13.5
  [53c48c17] + FixedPointNumbers v0.8.4
  [59287772] + Formatting v0.4.2
  [c91e804a] + Gadfly v1.3.4
  [42e2da0e] + Grisu v1.0.2
  [a1b4810d] + Hexagons v0.2.0
  [34004b35] + HypergeometricFunctions v0.3.11
  [09f84164] + HypothesisTests v0.10.11
  [9b13fd28] + IndirectArrays v1.0.0
  [842dd82b] + InlineStrings v1.3.2
  [a98d9a8b] + Interpolations v0.14.6
  [3587e190] + InverseFunctions v0.1.8
  [41ab1584] + InvertedIndices v1.1.0
  [92d709cd] + IrrationalConstants v0.1.1
  [c8e1da08] + IterTools v1.4.0
  [82899510] + IteratorInterfaceExtensions v1.0.0
  [692b3bcd] + JLLWrappers v1.4.1
  [682c06a0] + JSON v0.21.3
  [e5e0dc1b] + Juno v0.8.4
  [5ab0869b] + KernelDensity v0.6.5
  [b964fa9f] + LaTeXStrings v1.3.0
  [4345ca2d] + Loess v0.5.4
  [2ab3a3ac] + LogExpFunctions v0.3.19
  [1914dd2f] + MacroTools v0.5.10
  [442fdcdd] + Measures v0.3.2
  [e89f7d12] + Media v0.5.0
  [e1d29d7a] + Missings v1.0.2
  [77ba4419] + NaNMath v1.0.1
  [6fe1bfb0] + OffsetArrays v1.12.8
  [bac558e1] + OrderedCollections v1.4.1
  [90014a1f] + PDMats v0.11.16
  [69de0a69] + Parsers v2.5.1
  [2dfb63ee] + PooledArrays v1.4.2
  [21216c6a] + Preferences v1.3.0
  [08abe8d2] + PrettyTables v2.2.1
  [1fd47b50] + QuadGK v2.6.0
  [c84ed2f1] + Ratios v0.4.3
  [189a3867] + Reexport v1.2.2
  [ae029012] + Requires v1.3.0
  [79098fc4] + Rmath v0.7.0
  [f2b01f46] + Roots v2.0.8
  [91c51154] + SentinelArrays v1.3.16
  [efcf1570] + Setfield v1.1.1
  [992d4aef] + Showoff v1.0.3
  [66db9d55] + SnoopPrecompile v1.0.1
  [a2af1166] + SortingAlgorithms v1.1.0
  [276daf66] + SpecialFunctions v2.1.7
  [90137ffa] + StaticArrays v1.5.10
  [1e83bf80] + StaticArraysCore v1.4.0
  [82ae8749] + StatsAPI v1.5.0
  [2913bbd2] + StatsBase v0.33.21
  [4c63d2b9] + StatsFuns v1.0.1
  [69024149] + StringEncodings v0.3.5
  [892a3eda] + StringManipulation v0.3.0
  [3783bdb8] + TableTraits v1.0.1
  [bd369af6] + Tables v1.10.0
  [b718987f] + TextWrap v1.0.1
  [3bb67fe8] + TranscodingStreams v0.9.9
  [ea10d353] + WeakRefStrings v1.4.2
  [efce3f68] + WoodburyMatrices v0.5.5
  [ddb6d928] + YAML v0.4.8
  [f5851436] + FFTW_jll v3.3.10+0
  [1d5cc7b8] + IntelOpenMP_jll v2018.0.3+2
  [94ce4f54] + Libiconv_jll v1.16.1+1
  [856f044c] + MKL_jll v2022.2.0+0
  [efe28fd5] + OpenSpecFun_jll v0.5.5+0
  [f50d1b31] + Rmath_jll v0.3.0+0
  [0dad84c5] + ArgTools v1.1.1
  [56f22d72] + Artifacts
  [2a0f44e3] + Base64
  [ade2ca70] + Dates
  [8ba89e20] + Distributed
  [f43a241f] + Downloads v1.6.0
  [7b1f6079] + FileWatching
  [9fa8497b] + Future
  [b77e0a4c] + InteractiveUtils
  [4af54fe1] + LazyArtifacts
  [b27032c2] + LibCURL v0.6.3
  [76f85450] + LibGit2
  [8f399da3] + Libdl
  [37e2e46d] + LinearAlgebra
  [56ddb016] + Logging
  [d6f4376e] + Markdown
  [a63ad114] + Mmap
  [ca575930] + NetworkOptions v1.2.0
  [44cfe95a] + Pkg v1.8.0
  [de0858da] + Printf
  [9abbd945] + Profile
  [3fa0cd96] + REPL
  [9a3f8284] + Random
  [ea8e919c] + SHA v0.7.0
  [9e88b42a] + Serialization
  [1a1011a3] + SharedArrays
  [6462fe0b] + Sockets
  [2f01184e] + SparseArrays
  [10745b16] + Statistics
  [4607b0f0] + SuiteSparse
  [fa267f1f] + TOML v1.0.0
  [a4e569a6] + Tar v1.10.1
  [8dfed614] + Test
  [cf7118a7] + UUIDs
  [4ec0a83e] + Unicode
  [e66e0078] + CompilerSupportLibraries_jll v0.5.2+0
  [deac9b47] + LibCURL_jll v7.84.0+0
  [29816b5a] + LibSSH2_jll v1.10.2+0
  [c8ffd9c3] + MbedTLS_jll v2.28.0+0
  [14a3606d] + MozillaCACerts_jll v2022.2.1
  [4536629a] + OpenBLAS_jll v0.3.20+0
  [05823500] + OpenLibm_jll v0.8.1+0
  [83775a58] + Zlib_jll v1.2.12+3
  [8e850b90] + libblastrampoline_jll v5.1.1+0
  [8e850ede] + nghttp2_jll v1.48.0+0
  [3f19e933] + p7zip_jll v17.4.0+0
        Info Packages marked with ⌅ have new versions available but compatibility constraints restrict them from upgrading. To see why use `status --outdated -m`
[ Info: Loading simulation framework
[ Info: Directory test_output does not exist, attempting to create
[ Info: Using 1 thread(s)
[ Info: Parsing config
[ Info: Running config
[ Info: Generating plots
[ Info: Analyzing results
[ Info: Saving results in test_output

Quick results:
##############
Venn score = 0.991, 95% conf int (0.976, 1.006)
AUPRC score = 0.92, 95% conf int (0.895, 0.945)
SNR score = 3.706 +/- 0.371</code></pre><p>The <code>test_output/</code> directory should now be populated with all the files</p><pre class="documenter-example-output"><code class="nohighlight hljs ansi">counts.svg
results_table.csv
volcano.svg</code></pre><h2 id="Output"><a class="docs-heading-anchor" href="#Output">Output</a><a id="Output-1"></a><a class="docs-heading-anchor-permalink" href="#Output" title="Permalink"></a></h2><p>The folder contains one of the raw count scatterplots (left) and a volcano plot of mean log2 fold change versus significance of each gene (right)</p><p><img src="../test_output/counts.svg" alt/> <img src="../test_output/volcano.svg" alt/></p><p>It also has a useful table that contains all the summary statistic information.</p><div><div style = "float: left;"><span>10×8 DataFrame</span></div><div style = "clear: both;"></div></div><div class = "data-frame" style = "overflow-x: scroll;"><table class = "data-frame" style = "margin-bottom: 6px;"><thead><tr class = "header"><th class = "rowNumber" style = "font-weight: bold; text-align: right;">Row</th><th style = "text-align: left;">method</th><th style = "text-align: left;">measure</th><th style = "text-align: left;">genetype</th><th style = "text-align: left;">std_score</th><th style = "text-align: left;">mean_score</th><th style = "text-align: left;">conf_max</th><th style = "text-align: left;">conf_min</th><th style = "text-align: left;">n</th></tr><tr class = "subheader headerLastRow"><th class = "rowNumber" style = "font-weight: bold; text-align: right;"></th><th title = "InlineStrings.String7" style = "text-align: left;">String7</th><th title = "InlineStrings.String7" style = "text-align: left;">String7</th><th title = "InlineStrings.String15" style = "text-align: left;">String15</th><th title = "Float64" style = "text-align: left;">Float64</th><th title = "Float64" style = "text-align: left;">Float64</th><th title = "Float64" style = "text-align: left;">Float64</th><th title = "Float64" style = "text-align: left;">Float64</th><th title = "Int64" style = "text-align: left;">Int64</th></tr></thead><tbody><tr><td class = "rowNumber" style = "font-weight: bold; text-align: right;">1</td><td style = "text-align: left;">venn</td><td style = "text-align: left;">inc</td><td style = "text-align: left;">sigmoidal</td><td style = "text-align: right;">NaN</td><td style = "text-align: right;">NaN</td><td style = "text-align: right;">NaN</td><td style = "text-align: right;">NaN</td><td style = "text-align: right;">10</td></tr><tr><td class = "rowNumber" style = "font-weight: bold; text-align: right;">2</td><td style = "text-align: left;">auprc</td><td style = "text-align: left;">inc</td><td style = "text-align: left;">sigmoidal</td><td style = "text-align: right;">NaN</td><td style = "text-align: right;">NaN</td><td style = "text-align: right;">NaN</td><td style = "text-align: right;">NaN</td><td style = "text-align: right;">10</td></tr><tr><td class = "rowNumber" style = "font-weight: bold; text-align: right;">3</td><td style = "text-align: left;">venn</td><td style = "text-align: left;">dec</td><td style = "text-align: left;">sigmoidal</td><td style = "text-align: right;">0.0527046</td><td style = "text-align: right;">0.983333</td><td style = "text-align: right;">1.02633</td><td style = "text-align: right;">0.940333</td><td style = "text-align: right;">10</td></tr><tr><td class = "rowNumber" style = "font-weight: bold; text-align: right;">4</td><td style = "text-align: left;">auprc</td><td style = "text-align: left;">dec</td><td style = "text-align: left;">sigmoidal</td><td style = "text-align: right;">0.0453857</td><td style = "text-align: right;">0.955173</td><td style = "text-align: right;">0.992202</td><td style = "text-align: right;">0.918145</td><td style = "text-align: right;">10</td></tr><tr><td class = "rowNumber" style = "font-weight: bold; text-align: right;">5</td><td style = "text-align: left;">venn</td><td style = "text-align: left;">incdec</td><td style = "text-align: left;">sigmoidal</td><td style = "text-align: right;">0.0527046</td><td style = "text-align: right;">0.983333</td><td style = "text-align: right;">1.02633</td><td style = "text-align: right;">0.940333</td><td style = "text-align: right;">10</td></tr><tr><td class = "rowNumber" style = "font-weight: bold; text-align: right;">6</td><td style = "text-align: left;">auprc</td><td style = "text-align: left;">incdec</td><td style = "text-align: left;">sigmoidal</td><td style = "text-align: right;">0.0648829</td><td style = "text-align: right;">0.927664</td><td style = "text-align: right;">0.980599</td><td style = "text-align: right;">0.874728</td><td style = "text-align: right;">10</td></tr><tr><td class = "rowNumber" style = "font-weight: bold; text-align: right;">7</td><td style = "text-align: left;">venn</td><td style = "text-align: left;">inc</td><td style = "text-align: left;">linear</td><td style = "text-align: right;">NaN</td><td style = "text-align: right;">NaN</td><td style = "text-align: right;">NaN</td><td style = "text-align: right;">NaN</td><td style = "text-align: right;">10</td></tr><tr><td class = "rowNumber" style = "font-weight: bold; text-align: right;">8</td><td style = "text-align: left;">auprc</td><td style = "text-align: left;">inc</td><td style = "text-align: left;">linear</td><td style = "text-align: right;">NaN</td><td style = "text-align: right;">NaN</td><td style = "text-align: right;">NaN</td><td style = "text-align: right;">NaN</td><td style = "text-align: right;">10</td></tr><tr><td class = "rowNumber" style = "font-weight: bold; text-align: right;">9</td><td style = "text-align: left;">venn</td><td style = "text-align: left;">dec</td><td style = "text-align: left;">linear</td><td style = "text-align: right;">0.0</td><td style = "text-align: right;">1.0</td><td style = "text-align: right;">1.0</td><td style = "text-align: right;">1.0</td><td style = "text-align: right;">10</td></tr><tr><td class = "rowNumber" style = "font-weight: bold; text-align: right;">10</td><td style = "text-align: left;">auprc</td><td style = "text-align: left;">dec</td><td style = "text-align: left;">linear</td><td style = "text-align: right;">0.026003</td><td style = "text-align: right;">0.944467</td><td style = "text-align: right;">0.965682</td><td style = "text-align: right;">0.923252</td><td style = "text-align: right;">10</td></tr></tbody></table></div><p>The table below describes each column</p><table><tr><th style="text-align: left">Column Name</th><th style="text-align: left">Meaning</th></tr><tr><td style="text-align: left"><code>method</code></td><td style="text-align: left">Which summary statistic was used (e.g. <a href="../internals/#Crispulator.auprc"><code>Crispulator.auprc</code></a>)</td></tr><tr><td style="text-align: left"><code>measure</code></td><td style="text-align: left">Whether the score is only for increasing genes (<code>inc</code>), decreasing (<code>dec</code>) or both (<code>incdec</code>). Allows independent evaluation on which type of genes the screen can accurately evaluate.</td></tr><tr><td style="text-align: left"><code>genetype</code></td><td style="text-align: left">Whether the score is for linear, sigmoidal, or all genes (see <a href="../internals/#Crispulator.KDPhenotypeRelationship"><code>Crispulator.KDPhenotypeRelationship</code></a>). Helps determine if CRISPRn or CRISPRi is better for this design.</td></tr><tr><td style="text-align: left"><code>mean_score</code></td><td style="text-align: left">Average score</td></tr><tr><td style="text-align: left"><code>std_score</code></td><td style="text-align: left">Standard deviation in scores</td></tr><tr><td style="text-align: left"><code>conf_max</code></td><td style="text-align: left">Upper limit of 95% confidence interval</td></tr><tr><td style="text-align: left"><code>conf_min</code></td><td style="text-align: left">Lower limit of 95% confidence interval</td></tr><tr><td style="text-align: left"><code>n</code></td><td style="text-align: left">Number of independent replicates</td></tr></table></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../">« Home</a><a class="docs-footer-nextpage" href="../custom/">Custom Simulations »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 0.27.23 on <span class="colophon-date" title="Tuesday 22 November 2022 20:51">Tuesday 22 November 2022</span>. Using Julia version 1.8.3.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
