<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Tutorial · Crispulator.jl</title><script data-outdated-warner src="../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.045/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.13.24/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL=".."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../assets/documenter.js"></script><script src="../siteinfo.js"></script><script src="../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><a class="docs-logo" href="../"><img src="../assets/logo.png" alt="Crispulator.jl logo"/></a><div class="docs-package-name"><span class="docs-autofit"><a href="../">Crispulator.jl</a></span></div><form class="docs-search" action="../search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li><a class="tocitem" href="../">Home</a></li><li class="is-active"><a class="tocitem" href>Tutorial</a><ul class="internal"><li><a class="tocitem" href="#Graphical-overview"><span>Graphical overview</span></a></li><li><a class="tocitem" href="#Getting-started"><span>Getting started</span></a></li><li><a class="tocitem" href="#Running-simulation"><span>Running simulation</span></a></li><li><a class="tocitem" href="#Output"><span>Output</span></a></li></ul></li><li><a class="tocitem" href="../custom/">Custom Simulations</a></li><li><a class="tocitem" href="../internals/">Crispulator Internals</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>Tutorial</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Tutorial</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/tlnagy/Crispulator.jl/blob/master/docs/src/tutorial.md" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="Tutorial"><a class="docs-heading-anchor" href="#Tutorial">Tutorial</a><a id="Tutorial-1"></a><a class="docs-heading-anchor-permalink" href="#Tutorial" title="Permalink"></a></h1><p>The easiest way to use Crispulator is via the command line config file. If this format is too constraining, the <a href="../custom/#Custom-Simulations">Custom Simulations</a> has a detailed walk- through of writing a custom simulation where each step can be modified according to need.</p><h2 id="Graphical-overview"><a class="docs-heading-anchor" href="#Graphical-overview">Graphical overview</a><a id="Graphical-overview-1"></a><a class="docs-heading-anchor-permalink" href="#Graphical-overview" title="Permalink"></a></h2><p>The simulation is laid out in the following manner:</p><p><img src="../assets/overview.png" alt/></p><h2 id="Getting-started"><a class="docs-heading-anchor" href="#Getting-started">Getting started</a><a id="Getting-started-1"></a><a class="docs-heading-anchor-permalink" href="#Getting-started" title="Permalink"></a></h2><p>First, navigate to the Crispulator directory.</p><div class="admonition is-success"><header class="admonition-header">Tip</header><div class="admonition-body"><p>You can find the directory by running</p><pre><code class="language-sh hljs">$ julia -e &#39;println(Pkg.dir(&quot;Crispulator&quot;))&#39;</code></pre></div></div><p>There should be a <a href="https://learnxinyminutes.com/docs/yaml/">YAML file</a> called <code>example_config.yml</code>. Open this is in a text editor and it should look like this</p><pre><code class="language-yaml hljs"># This is an example configuration file. Whitespace is important.

# Settings pertaining to the library design
library:
    genome:
        num-genes: 500
        num-guides-per-gene: 5
        frac-increasing-genes: 0.02 # fraction of genes with a positive phenotype
        frac-decreasing-genes: 0.1 # fraction of genes with a negative phenotype

    guides:
        crispr-type: CRISPRn # either CRISPRi or CRISPRn
        frac-high-quality: 0.9 # fraction of high quality guides
        mean-high-quality-kd: 0.85 # mean knockdown by a high quality guide (CRISPRi only)

screen:
    type: facs # either facs or growth
    num-runs: 10 # how many independent runs

    representation: # integer value, how much larger are samples than the library
        - transfection: 100
        - selection: 100
        - sequencing: 100

# screen-type specific parameters

    bin-size: 0.25 # size of tail to sample from, must be between 0 and 0.5 (FACS only)
    std-noise: 1 # (FACS only)
    num-bottlenecks: 10 # (Growth only)</code></pre><p>This gives access to most dials in the simulation, if something is missing than see <a href="../custom/#Custom-Simulations">Custom Simulations</a>.</p><p>Now, lets remove all genes that have a positive phenotype by changing line 8 to 0.0:</p><pre><code class="language-yaml hljs">        frac-increasing-genes: 0.0 # fraction of genes with a positive phenotype</code></pre><h2 id="Running-simulation"><a class="docs-heading-anchor" href="#Running-simulation">Running simulation</a><a id="Running-simulation-1"></a><a class="docs-heading-anchor-permalink" href="#Running-simulation" title="Permalink"></a></h2><p>Now, we can actually run the code by executing the following command</p><pre><code class="language-julia hljs">julia run.jl config example_config.yml test_output</code></pre><div class="admonition is-success"><header class="admonition-header">Tip</header><div class="admonition-body"><p>Here <code>config</code> tells <code>CRISPulator</code> to use the provided config <code>example_config.yml</code> and <code>test_output</code> is the directory where the results will be saved. This directory will be created if it doesn&#39;t exist.</p></div></div><p>The output should look like</p><pre><code class="language- hljs">output = read(`julia ../../run.jl config example_config.yml test_output`, String) # hide
println(output) # hide</code></pre><p>The <code>test_output/</code> directory should now be populated with all the files</p><pre><code class="language- hljs">foreach(println, readdir(&quot;test_output&quot;)) # hide</code></pre><h2 id="Output"><a class="docs-heading-anchor" href="#Output">Output</a><a id="Output-1"></a><a class="docs-heading-anchor-permalink" href="#Output" title="Permalink"></a></h2><p>The folder contains one of the raw count scatterplots (left) and a volcano plot of mean log2 fold change versus significance of each gene (right)</p><p><img src="test_output/counts.svg" alt/> <img src="test_output/volcano.svg" alt/></p><p>It also has a useful table that contains all the summary statistic information.</p><pre><code class="language- hljs">using DataFrames # hide
using CSV # hide
first(CSV.read(joinpath(&quot;test_output&quot;, &quot;results_table.csv&quot;), DataFrame), 10) # hide</code></pre><p>The table below describes each column</p><table><tr><th style="text-align: left">Column Name</th><th style="text-align: left">Meaning</th></tr><tr><td style="text-align: left"><code>method</code></td><td style="text-align: left">Which summary statistic was used (e.g. <a href="../internals/#Crispulator.auprc"><code>Crispulator.auprc</code></a>)</td></tr><tr><td style="text-align: left"><code>measure</code></td><td style="text-align: left">Whether the score is only for increasing genes (<code>inc</code>), decreasing (<code>dec</code>) or both (<code>incdec</code>). Allows independent evaluation on which type of genes the screen can accurately evaluate.</td></tr><tr><td style="text-align: left"><code>genetype</code></td><td style="text-align: left">Whether the score is for linear, sigmoidal, or all genes (see <a href="../internals/#Crispulator.KDPhenotypeRelationship"><code>Crispulator.KDPhenotypeRelationship</code></a>). Helps determine if CRISPRn or CRISPRi is better for this design.</td></tr><tr><td style="text-align: left"><code>mean_score</code></td><td style="text-align: left">Average score</td></tr><tr><td style="text-align: left"><code>std_score</code></td><td style="text-align: left">Standard deviation in scores</td></tr><tr><td style="text-align: left"><code>conf_max</code></td><td style="text-align: left">Upper limit of 95% confidence interval</td></tr><tr><td style="text-align: left"><code>conf_min</code></td><td style="text-align: left">Lower limit of 95% confidence interval</td></tr><tr><td style="text-align: left"><code>n</code></td><td style="text-align: left">Number of independent replicates</td></tr></table></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../">« Home</a><a class="docs-footer-nextpage" href="../custom/">Custom Simulations »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 0.27.23 on <span class="colophon-date" title="Tuesday 22 November 2022 08:33">Tuesday 22 November 2022</span>. Using Julia version 1.8.3.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
